src/isomorphic/children/ReactChildren.js

```diff
@@ -127,58 +196,48 @@ function mapSingleChildIntoContext(bookKeeping, child, childKey) {
   }
 }
 
-function mapIntoWithKeyPrefixInternal(children, array, prefix, func, context) {
-  var escapedPrefix = '';
-  if (prefix != null) {
-    escapedPrefix = escapeUserProvidedKey(prefix) + '/';
-  }
-  var traverseContext = MapBookKeeping.getPooled(
-    array,
-    escapedPrefix,
-    func,
-    context,
-  );
-  traverseAllChildren(children, mapSingleChildIntoContext, traverseContext);
-  MapBookKeeping.release(traverseContext);
-}
-
 /**
  * Maps children that are typically specified as `props.children`.
  *
  * See https://facebook.github.io/react/docs/react-api.html#react.children.map
  *
  * The provided mapFunction(child, key, index) will be called for each
  * leaf child.
- *
- * @param {?*} children Children tree container.
- * @param {function(*, int)} func The map function.
- * @param {*} context Context for mapFunction.
- * @return {object} Object containing the ordered map of results.
  */
-function mapChildren(children, func, context) {
+function mapChildren(
+  children: mixed,
+  func: () => mixed,
+  context?: Object,
+): ?(mixed[]) {
   if (children == null) {
     return children;
   }
   var result = [];
-  mapIntoWithKeyPrefixInternal(children, result, null, func, context);
+  var traverseContext = {
+    result: result,
+    keyPrefix: '',
+    func: func,
+    context: context,
+    count: 0,
+  };
+  traverseAllChildren(children, '', mapSingleChildIntoContext, traverseContext);
   return result;
 }
 
-function forEachSingleChildDummy(traverseContext, child, name) {
-  return null;
-}
-
 /**
- * Count the number of children that are typically specified as
- * `props.children`.
+ * Iterates through children that are typically specified as `props.children`.
  *
- * See https://facebook.github.io/react/docs/react-api.html#react.children.count
+ * See https://facebook.github.io/react/docs/react-api.html#react.children.foreach
  *
- * @param {?*} children Children tree container.
- * @return {number} The number of children.
+ * The provided forEachFunc(child, index) will be called for each
+ * leaf child.
  */
-function countChildren(children, context) {
-  return traverseAllChildren(children, forEachSingleChildDummy, null);
+function forEachChildren(
+  children: mixed,
+  forEachFunc: () => mixed,
+  forEachContext?: Object,
+): void {
+  mapChildren(children, forEachFunc, forEachContext);
```

Comment:
We don't want to implement `forEach` in terms of `map` because we don't want to allocate the result just to ignore it. It's fine to duplicate the code in this case.